python/cuml/cuml/linear_model/linear_regression.pyx

```diff
@@ -306,210 +273,214 @@ class LinearRegression(Base,
             **super()._attrs_to_cpu(model),
         }
 
-    def __init__(self, *, algorithm='eig', fit_intercept=True,
-                 copy_X=True, normalize=False,
-                 handle=None, verbose=False, output_type=None):
-        if handle is None and algorithm == 'eig':
+    def __init__(
+        self,
+        *,
+        algorithm="auto",
+        fit_intercept=True,
+        copy_X=True,
+        normalize=False,
+        handle=None,
+        verbose=False,
+        output_type=None
+    ):
+        if handle is None and algorithm in ("auto", "eig"):
             # if possible, create two streams, so that eigenvalue decomposition
             # can benefit from running independent operations concurrently.
             handle = Handle(n_streams=2)
-        super().__init__(handle=handle,
-                         verbose=verbose,
-                         output_type=output_type)
 
-        # internal array attributes
-        self.coef_ = None
-        self.intercept_ = None
+        super().__init__(handle=handle, verbose=verbose, output_type=output_type)
 
+        self.algorithm = algorithm
         self.fit_intercept = fit_intercept
+        self.copy_X = copy_X
         self.normalize = normalize
-        if algorithm in ['svd', 'eig', 'qr', 'svd-qr', 'svd-jacobi']:
-            self.algorithm = algorithm
-            self.algo = self._get_algorithm_int(algorithm)
-        else:
-            msg = "algorithm {!r} is not supported"
-            raise TypeError(msg.format(algorithm))
 
-        self.intercept_value = 0.0
-        self.copy_X = copy_X
+    def _select_algo(self, X, y):
+        """Select the solver algorithm based on `algorithm` and problem dimensions"""
+        if X.shape[0] == 1:
+            fallback_reason = "single-column X"
+        elif y.ndim == 2 and y.shape[1] > 1:
+            fallback_reason = "multi-column y"
+        else:
+            fallback_reason = None
 
-    def _get_algorithm_int(self, algorithm):
-        return {
-            'svd': 0,
-            'eig': 1,
-            'qr': 2,
-            'svd-qr': 3,
-            'svd-jacobi': 0
-        }[algorithm]
+        if self.algorithm == "auto":
+            algo = Algo.SVD if fallback_reason else Algo.EIG
+        else:
+            algo = Algo.parse(self.algorithm)
+            if fallback_reason and algo != Algo.SVD:
+                warnings.warn(
+                    (
+                        "Falling back to `algorithm='svd'` as `algorithm="
+                        "{self.algorithm!r}` doesn't support {fallback_reason}."
+                    ),
+                    UserWarning,
+                )
+                algo = Algo.SVD
```

Comment:
Wouldn't it be better to raise an error and let the user restart with appropriate parameters?

Reply:
If we change that behavior we should do it across all estimators IMO (not sure if Ridge would be the only one).